daemon/network.go

```diff
@@ -99,15 +100,51 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+	ingress            sync.Mutex
```

Comment:
Thanks for adding the lock. But `startIngressWorker` is using `ingressWorkerStop` without holding the lock, and `SetupIngress` / `ReleaseIngress` are using `ingressWorkerOnce` without holding the lock, so unfortunately my concerns about concurrency are not fully addressed.

Earlier in the thread you said:

> The fact is the `SetupIngressNetwork` being called from the executor and the
call to `DaemonLeavesCluster` happen from two different threads. Their execution
order is not guaranteed, and I saw it not deterministic during the testing of
initial diffs.
> 
> In fact in the setup ingress call, we need to wait for `agentInitWait()` for
this reason.
> 
> So it is not excluded a `swarm join` quickly followed by a `swarm leave` would
flip the order of events.

...so I still don't see how it's safe to use these variables from different goroutines without a lock.